### PR TITLE
:sparkles: Move ID from Entity to curate

### LINKED ISSUE #243: Why is the reference field a class arg and not an arg of `curate()`?
I'm fine the current design, but I think it'd feel more natural if `id` was an arg of `curate` and had a more explicit name.

However, I understand that if several class behaviors are affected (not just curate), any `CellType` instance should have that `id` field customized.

image

### PR DESCRIPTION
**Pull request by Zethson**
Fix https://github.com/laminlabs/bionty/issues/243

1. Removed ID parameter from all entities.
2. Added a "reference_index" parameter to curate that defines the reference column/index of the ontology to curate against
3. Renamed "column" to "target_column" of curate
4. Adapted the docs accordingly.

### CODE REVIEW

**Comment by Zethson** [comment]
I'll improve ID handling soon in this PR

**Comment by Zethson** [comment]
This PR should be against staging. Will change.

**Comment by sunnyosun** [comment]
Could you release v0.11.0 after merging this? Then we can properly start a staging branch 0.12?

**Comment by Zethson** [comment]
> Could you release v0.11.0 after merging this? Then we can properly start a staging branch 0.12?

I can't make releases because I don't have the credentials, right?

**Comment by Zethson** [comment]
@sunnyosun see PR description.

Was wondering whether we should name it "reference_index" or "reference_column". Thoughts?

Hope that I elsewise implemented what we desired.

**Comment by sunnyosun** [comment]
> > Could you release v0.11.0 after merging this? Then we can properly start a staging branch 0.12?
> 
> I can't make releases because I don't have the credentials, right?

I'll give you permission! see slack~

**Comment by sunnyosun** [comment]
> @sunnyosun see PR description.
> 
> Was wondering whether we should name it "reference_index" or "reference_column". Thoughts?
> 
> Hope that I elsewise implemented what we desired.

Could we call it `reference_id`? I'm thinking of it as in an id conversion tool, passing identifiers and mapping them against a `reference_id`. Users don't need to know it's the index of the reference dataframe I think, that's just under the hood for us.

We can even in the future introduce a `.map` function which takes a list of identifiers and returns the converted ones.

File: bionty/_entity.py [hunk 1]
```diff
@@ -37,20 +37,22 @@ class Entity:
     def __init__(
         self,
         database: Optional[str],
-        id: Optional[str] = None,
         version: Optional[str] = None,
         species: Optional[str] = None,
         filenames: Optional[Dict[str, str]] = None,
         *,
         prefix: Optional[str] = None,
+        reference_index: Optional[str] = None,
```
Comment (sunnyosun) [inline bionty/_entity.py:45]: I think we can just remove this?

File: bionty/_entity.py [hunk 2]
```diff
@@ -122,7 +124,9 @@ def df(self) -> pd.DataFrame:
         df.to_parquet(self._local_parquet_path)
 
         return (
-            pd.read_parquet(self._local_parquet_path).reset_index().set_index(self._id)
+            pd.read_parquet(self._local_parquet_path)
+            .reset_index()
+            .set_index(self.reference_index)
```
Comment (sunnyosun) [inline bionty/_entity.py:129]: Could we remove `.set_index(self.reference_index)` here? Then you no longer need this parameter from init.

File: bionty/_entity.py [hunk 3]
```diff
@@ -287,36 +291,55 @@ def _set_attributes(
         )
 
     def curate(
-        self, df: pd.DataFrame, column: str = None, case_sensitive: bool = True
+        self,
+        df: pd.DataFrame,
+        reference_index: str = "ontology_id",
+        target_column: str = None,
```
Comment (sunnyosun) [inline bionty/_entity.py:297]: Could we rename it back to `column`? I think from user side, they should just think of a single table, that's their input dataframe, and they want to curate a column in the dataframe. There is no confusion around target or not as they won't see the underlying process of using the reference table for curation.

Comment (sunnyosun) [reply bionty/_entity.py:297]: Could we have the `column` as the 2nd parameter? As it's more related to the 1st parameter df?

File: bionty/_entity.py [hunk 4]
```diff
@@ -37,20 +37,22 @@ class Entity:
...
         species: Optional[str] = None,
         filenames: Optional[Dict[str, str]] = None,
         *,
         prefix: Optional[str] = None,
+        reference_index: Optional[str] = None,
     ):
-        self._id = "id" if id is None else id
         # By default lookup allows auto-completion for the `name` field.
         # lookup column can be changed using `.lookup_col = `.
         self._lookup_col = "name"
```
Comment (sunnyosun) [inline bionty/_entity.py:49]: In this PR, could you also rename `.lookup_col` to `.lookup_field`? This is for configuring which id to use for lookup auto-completion, it defaults to "name" but can be set after the class is initialized.

**Review by sunnyosun — APPROVED** [review]
Thank you, Lukas, this is great! 🙏

File: bionty/_entity.py (hunk 1 above)
Comment (Zethson) [reply bionty/_entity.py:45]: Not quite. The default reference isn't "ontology_id" for all ontologies. This allows us to pass default references for a few ontologies where it makes sense (prettier DataFrame printing). This is currently a positional only parameter so I'd keep it.

File: bionty/_entity.py (hunk 2 above)
Comment (Zethson) [reply bionty/_entity.py:129]: That's what I had at some moment, but this allows for prettier DF printing

**Pull request merged**